cinder/controllerserver.go

```diff
@@ -0,0 +1,193 @@
+			},
+		},
+	}, nil
+}
+
+func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+
+	// Get OpenStack Provider
+	cloud, err := openstack.GetOpenStackProvider()
+	if err != nil {
+		glog.V(3).Infof("Failed to GetOpenStackProvider: %v", err)
+		return nil, err
+	}
+
+	// Volume Delete
+	volID := req.GetVolumeId()
+	err = cloud.DeleteVolume(volID)
+	if err != nil {
+		glog.V(3).Infof("Failed to CreateVolume: %v", err)
+		return nil, err
+	}
+
+	glog.V(4).Infof("Delete volume %s", volID)
+
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+func (cs *controllerServer) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+
+	// Get OpenStack Provider
+	cloud, err := openstack.GetOpenStackProvider()
+	if err != nil {
+		glog.V(3).Infof("Failed to GetOpenStackProvider: %v", err)
+		return nil, err
+	}
+
+	// Volume Attach
+	instanceID := req.GetNodeId()
+	volumeID := req.GetVolumeId()
+
+	_, err = cloud.AttachVolume(instanceID, volumeID)
+	if err != nil {
+		glog.V(3).Infof("Failed to AttachVolume: %v", err)
+		return nil, err
+	}
+
+	err = cloud.WaitDiskAttached(instanceID, volumeID)
+	if err != nil {
+		glog.V(3).Infof("Failed to WaitDiskAttached: %v", err)
+		return nil, err
+	}
+
+	devicePath, err := cloud.GetAttachmentDiskPath(instanceID, volumeID)
+	if err != nil {
+		glog.V(3).Infof("Failed to GetAttachmentDiskPath: %v", err)
+		return nil, err
+	}
+
+	glog.V(4).Infof("ControllerPublishVolume %s on %s", volumeID, instanceID)
+
+	// Publish Volume Info
+	pvInfo := map[string]string{}
+	pvInfo["DevicePath"] = devicePath
+
+	return &csi.ControllerPublishVolumeResponse{
+		PublishVolumeInfo: pvInfo,
+	}, nil
+}
+
+func (cs *controllerServer) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+
+	// Get OpenStack Provider
+	cloud, err := openstack.GetOpenStackProvider()
+	if err != nil {
+		glog.V(3).Infof("Failed to GetOpenStackProvider: %v", err)
+		return nil, err
+	}
+
+	// Volume Detach
+	instanceID := req.GetNodeId()
+	volumeID := req.GetVolumeId()
+
+	err = cloud.DetachVolume(instanceID, volumeID)
+	if err != nil {
+		glog.V(3).Infof("Failed to DetachVolume: %v", err)
+		return nil, err
+	}
+
+	err = cloud.WaitDiskDetached(instanceID, volumeID)
+	if err != nil {
+		glog.V(3).Infof("Failed to WaitDiskDetached: %v", err)
+		return nil, err
+	}
+
+	glog.V(4).Infof("ControllerUnpublishVolume %s on %s", volumeID, instanceID)
+
+	return &csi.ControllerUnpublishVolumeResponse{}, nil
+}
+
+func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+	for _, cap := range req.GetVolumeCapabilities() {
+		if cap.GetAccessMode().GetMode() != csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER {
```

Comment:
We do not need to implement this. I have fixed the default ValidateVolumeCapabilities, which loops over csiDriver caps.
https://github.com/kubernetes-csi/drivers/blob/master/csi-common/controllerserver-default.go#L47

Reply:
Ok,  I will remove it.